### DISCUSSION TITLE
Add ignite event filters to monai handlers (StatsHandler in particular)

### DISCUSSION THREAD

**Discussion by jak0bw** [Project-MONAI/MONAI#5958] [Ideas]
I am sorry if I missed an already existing feature but I could not find an option to use the event filter option (https://pytorch.org/ignite/generated/ignite.engine.events.Events.html paragraphs starts with Since v0.3.0) for `monai.handler.StatsHandler`.

Originally I only wanted to log the stats for every Xth iteration but wanted to use the default logger. I solved this with a lambda function which refers to to the created `StatsHandler`:
`
stats_handler = StatsHandler(iteration_print_logger=lambda engine: stats_handler._default_iteration_print(engine)
        if engine.state.iteration % iteration_log_interval == 0 else None,
        tag_name='Mean Squared Error',
        output_transform=from_engine(['loss'], first=True))
`

But this feels a little clunky to me. As a solution I had the idea to expose the ignite.event_filter feature to the user by adding an `iteration_filter` and an `epoch_filter` argument to the `StatsHandler` Class. These would enable setting specifying event filters for both the `ITERATION_COMPLETED` and `EPOCH_COMPLETED`  events in the attach function https://github.com/Project-MONAI/MONAI/blob/cdac0330b7991ca0ba3807f7c3873a0d3b2588bc/monai/handlers/stats_handler.py#L122-L142 (if `iteration_log` or `epoch_log` is 'True'). It is currently possible to solve this problem by writing a new logger function (more or less what I did above) and handle the event_filtering there but to me it feels like exposing the ignite.event_filter feature is the better solution to keep the logging and filter functionality separated. 

This idea of exposing the ignite.event_filter feature to the user might also apply to other `monai.handlers`.

**Comment by Nic-Ma** [comment]
Thanks for raising it.
CC @vfdev-5 , any suggestion for this feature request?

Thanks in advance.

**Reply by vfdev-5** [reply]
@jak0bw yes, this could be an interesting feature. I think we can even do something more cool by overriding `iteration_log` and `epoch_log` args. Currently, they are booleans but we could extend them to `Union[bool, Callable]` where if arg is callable then it could be used for event filtering:

```python
    def attach(self, engine):


         if self.iteration_log and not engine.has_event_handler(self.iteration_completed, Events.ITERATION_COMPLETED):
             event_filter = self.iteration_log if callable(self.iteration_log) else None
             engine.add_event_handler(
                 Events.ITERATION_COMPLETED(event_filter=event_filter), self.iteration_completed
             ) 
         if self.epoch_log and not engine.has_event_handler(self.epoch_completed, Events.EPOCH_COMPLETED):
             event_filter = self.epoch_log if callable(self.epoch_log) else None
             engine.add_event_handler(
                 Events.EPOCH_COMPLETED(event_filter=event_filter), self.epoch_completed
             ) 
```

**Reply by Nic-Ma** [reply]
Hi @vfdev-5 ,

Your proposal sounds interesting to me.
@wyli @ericspod What do you guys think?

Thanks.

**Reply by wyli** [reply]
agreed, I'm creating a feature request from this... 

(updated: https://github.com/Project-MONAI/MONAI/issues/5964)

**Reply by jak0bw** [reply]
Looks good to me as well. Just wanted to add two things.

1) Is it desirable two emulate the full functionality of event filters? Currently I believe the proposal only applies to option a) custom event filters and ignores b) every and c) once which look quite convenient to me (but I do currently do not have a great idea how to include them here, as they are optional arguments during event creation... -> pytorch ignite source code: https://pytorch.org/ignite/_modules/ignite/engine/events.html)
From https://pytorch.org/ignite/generated/ignite.engine.events.Events.html:
![image](https://user-images.githubusercontent.com/65398308/217809514-a68b87af-653a-4942-afaf-9da4098360cd.png)

2) Is it desirable to extend (some of) this functionality to other monai handlers such as
- Validation Handler https://docs.monai.io/en/stable/handlers.html#validation-handler
- Parameter Scheduler handler https://docs.monai.io/en/stable/handlers.html#parameter-scheduler-handler
- etc.

**Reply by vfdev-5** [reply]
Thanks @jak0bw. I was also thinking about that in the PR : https://github.com/Project-MONAI/MONAI/pull/5965

As for the point 1. To include `every` option, we may further extend `iteration_log` type to `bool | Callable | int`. I do not know and not sure if `once` option makes sense for `StatsHandler`, what do you think @jak0bw ?

**Reply by jak0bw** [reply]
@vfdev-5 I was kind of hoping for a systematic solution that covers all three options and could be applied to all affected handlers, but as I said I do not really have an idea how to do this elegantly.

Under these circumstances I think your suggestions to extend `iteration_log` to `bool | Callable | int` sounds like quite a good idea. Ignoring `once` should probably cover most instances and in case someone needs `once`, a lambda function has to be used...?

**Reply by vfdev-5** [reply]
>  Ignoring once should probably cover most instances and in case someone needs once, a lambda function has to be used...?

yes, `iteration_log=lambda e, i: i == 12` should probably do the same as "once=12".

Just an idea: maybe, we could also parse a string value, like `iteration_log="every=12"` or `iteration_log="once=12"` but such approach may look a bit strange and can be more error prone during the parsing part...

**Reply by jak0bw** [reply]
I am personally not a big fan of pure string parsing but maybe something like changing the data_type of `iteration_log` to `bool | Tuple[ event_filter_enum, callable | int]` and event_filter_enum being an `enum` of `event_filter | once | every `. Then it can be `bool` for back compatibility but also all event filter options are supported and we avoid pure string parsing?

**Reply by vfdev-5** [reply]
By the way, we haven't yet released that in stable but on master we have more options: every, once, before, after.
https://github.com/pytorch/ignite/blob/d2020e4e253ac1455a757c2db895c68ccfd2b958/ignite/engine/events.py#L62-L68

I agree that if we would like to have full support of those args, it would be nice to have a way to distinguish between the options.
Maybe, 
```
iteration_log: bool | Callable | Dict[str, int]
```
Such that `Callable` is directly corresponds to `event_filter` and dicts can be: `{"every": N}` or `{"once", N}` or `{"before", N}` or `{"after", N}` and also `{"after": M, "before": N}`...

**Reply by jak0bw** [reply]
> By the way, we haven't yet released that in stable but on master we have more options: every, once, before, after. https://github.com/pytorch/ignite/blob/d2020e4e253ac1455a757c2db895c68ccfd2b958/ignite/engine/events.py#L62-L68

Ah interesting, I didn't know that but sounds quite interesting and I think we would want to support it fully...

> ```
> iteration_log: bool | Callable | Dict[str, int]
> ```
> 
> Such that `Callable` is directly corresponds to `event_filter` and dicts can be: `{"every": N}` or `{"once", N}` or `{"before", N}` or `{"after", N}` and also `{"after": M, "before": N}`...


This then looks like quite a sensible solution, I think.

Just for my understanding after this feature is merged to stable, it will be possible in poytorch-ignite to mix multiple arguments for example after=9, before=21 and every=5, to only execute for iterations [10, 15, 20]?

**Reply by vfdev-5** [reply]
> Just for my understanding after this feature is merged to stable, it will be possible in poytorch-ignite to mix multiple arguments for example after=9, before=21 and every=5, to only execute for iterations [10, 15, 20]?

Right now, only`before` and `after` can work together but not with `every` or `once`. For complex filtering user should use event_filter function. 
However, your example "after=9, before=21 and every=5, to only execute for iterations [10, 15, 20]" looks good to support as well. We have to think about.

**Reply by jak0bw** [reply]
Just my two cents on this topic. I think being able to use `before`, `after`, and `every` together (and keeping `once` separately) makes sense as it probably covers most instances and ist not that complex (in my opinion at least). The custom event filter could then be mostly used to filter on engine state and only in obscure instances to filter on the iteration counter, if this makes sense?

I only could think of one other (maybe) useful option in extending `once` to `int | list[int]` to enable giving a list of `[5,8,23]` to only execute once for iterations 5, 8 and 23 (if this isn't already supported).

**Reply by vfdev-5** [reply]
> Just my two cents on this topic. I think being able to use before, after, and every together (and keeping once separately) makes sense as it probably covers most instances and ist not that complex (in my opinion at least).
The custom event filter could then be mostly used to filter on engine state and only in obscure instances to filter on the iteration counter, if this makes sense?

I agree

> I only could think of one other (maybe) useful option in extending once to int | list[int] to enable giving a list of [5,8,23] to only execute once for iterations 5, 8 and 23 (if this isn't already supported).

Thanks for the suggestion, seems interesting to support !

(Created new issue for that : https://github.com/pytorch/ignite/issues/2853)

**Reply by jak0bw** [reply]
Thanks for listening, love what you do 🤗
